discovery_child_development/notebooks/openalex/06_fine_tune_embeddings.py

```diff
@@ -214,6 +217,14 @@ def load_trainer(
     )
 
 
+# %%
+run = wandb.init(
+    project="ISS supervised ML",
+    job_type="Taxonomy classifier",
+    save_code=True,
+    tags=["finetuning"],
+)
+
 # %%
 # Load the data. Just the training set by default
 openalex_data, training_file_name = oa.get_labelled_data()
```

Comment:
How do the two different types of embeddings work? You load in 'embeddings' and join that into the openalex data. But then also use df_to_hf_ds, which also tokenizes the data? Is the second one used in the model?

Reply:
Yes good spot! Loading in the embeddings was not needed. I think the tokenizing step creates the embeddings that are needed as input to the Distilbert training.

Reply:
Good question re: num_labels. I did not know the answer. [This tutorial](https://medium.com/analytics-vidhya/finetune-distilbert-for-multi-label-text-classsification-task-994eb448f94c) also supplies a parameter called `num_classes` so I think it is mandatory. Have you seen anything that makes you think there's a default?

I asked ChatGPT and it had a whole lot to say on the topic!
"""
In the HuggingFace Transformers library, the num_labels parameter in the AutoModelForSequenceClassification function is used to specify the number of distinct labels that the model should predict for a given task. This is particularly relevant for classification tasks, where the model needs to predict which category or class a given input belongs to.

When you're dealing with a classification problem, the model outputs a prediction for each label, and the num_labels parameter tells the model how many different output labels it needs to prepare for. This is crucial for the model to understand the shape and structure of the output layer, where the final classification decisions are made.

For instance, if you are training a sentiment analysis model where the input texts can be classified as either positive, negative, or neutral, you would set num_labels=3 since there are three possible categories.

In the context of the code snippet you provided, where problem_type is set to "multi_label_classification", the num_labels parameter becomes even more significant. It indicates that each instance in the dataset can belong to multiple categories at once, and the model needs to predict a set of labels, not just one. Each label in the output will be treated as a separate binary classification, determining whether or not the instance belongs to that label.
"""

Reply:
I love that you asked chatgpt ahaha. I was having a look at the [source code](https://github.com/huggingface/transformers/blob/v4.35.2/src/transformers/configuration_utils.py#L49). You can just search for num_labels and it seems that if you don't specify if will go to 2? Which is definitely not what you wanted (but is what i want), so I think you have done it correctly! I think I may have just been misunderstanding what  `len(train_ds[0]["labels"])` was returning. Is that the maximum number of categories possible?